File: src/BIOME4Py/growth_subroutines/fire.py
"""Calculate the number of potential fire days
in a year based on threshold values for soil moisture, which are
PFT dependent.  Jed Kaplan 1998."""

# Standard library
import math
from dataclasses import dataclass
from typing import List


@dataclass
class FireResult:
    firedays: float
    wetday: float
    dryday: float
    firefraction: float
    burnfraction: float


def fire(wet: List[float], pft: int, lai: float, npp: float) -> FireResult:
    threshold = [
        0.25,
        0.20,
        0.40,
        0.33,
        0.40,
        0.33,
        0.33,
        0.40,
        0.40,
        0.33,
        0.33,
        0.33,
        0.33,
    ]
    firedays = 0.0
    wetday = 0.0
    dryday = 100.0
    burn = [0.0] * 365

    for day in range(365):
        if wet[day] < threshold[pft]:
            burn[day] = 1.0
        elif wet[day] > threshold[pft] + 0.05:
            burn[day] = 0.0
        else:
            burn[day] = 1 / math.exp(wet[day] - threshold[pft])

        if wet[day] > wetday:
            wetday = wet[day]
        if wet[day] < dryday:
            dryday = wet[day]

        firedays += burn[day]

    firefraction = firedays / 365.0
    litter = (lai / 5.0) * npp
    burnfraction = litter * (1 - (math.exp(-0.2 * firefraction**1.5)) ** 1.5)

    if npp < 1000.0:
        firedays *= npp / 1000.0

    return FireResult(
        firedays=firedays,
        wetday=wetday,
        dryday=dryday,
        firefraction=firefraction,
        burnfraction=burnfraction,
    )


if __name__ == "__main__":
    fire()

File: src/models/MechanisticModel/growth_subroutines/fire.jl

"""
  fire(wet::AbstractVector{T}, pft::U, lai::T, npp::T)::AbstractPFT

Calculate the number of potential fire days in a year based on threshold values for soil moisture.

# Arguments
- `wet`: A vector of 365 T values representing daily wetness.
- `pft`: An integer representing the Plant Functional Type (PFT).
- `lai`: The Leaf Area Index.
- `npp`: Net Primary Productivity.

# Returns
- 'pft`: An updated Plant Functional Type (PFT) with the number of fire days set.'
"""
function fire(wet::AbstractVector{T},
    pft::AbstractPFT,
    lai::T,
    npp::T
)::AbstractPFT where {T <: Real, U <: Int}

    # Threshold values per PFT
    threshold = get_characteristic(pft, :threshold)

    # Initialize variables
    firedays = T(0.0)
    wetday = T(0.0)
    dryday = T(100.0)
    burn = zeros(T, 365)

    # Loop through each day of the year
    for day in 1:365
        if wet[day] < threshold
            burn[day] = T(1.0)
        elseif wet[day] > threshold + T(0.05)
            burn[day] = T(0.0)
        else
            burn[day] = T(1.0) / exp(wet[day] - threshold)
        end

        # Update wetday and dryday
        wetday = max(wetday, wet[day])
        dryday = min(dryday, wet[day])

        # Accumulate firedays
        firedays += burn[day]
    end

    # Calculate fire fraction and burn fraction
    firefraction = firedays / T(365.0)
    litter = (lai / T(5.0)) * npp
    burnfraction = litter * (T(1.0) - (exp(-T(0.2) * firefraction^T(1.5)))^T(1.5))

    # Adjust firedays based on NPP
    if npp < T(1000.0)
        firedays *= npp / T(1000.0)
    end

    set_characteristic(pft, :firedays, firedays)

    # Return results as a FireResult
    return pft
end